examples/general/non_local_effect_before_error_return/src/lib.rs

```diff
@@ -221,9 +221,22 @@ fn in_async_function(tcx: ty::TyCtxt<'_>, hir_id: rustc_hir::HirId) -> bool {
         })
 }
 
-fn is_result(cx: &LateContext<'_>, ty: ty::Ty) -> bool {
-    if let ty::Adt(adt, _) = ty.kind() {
-        cx.tcx.is_diagnostic_item(sym::Result, adt.did())
+fn is_permitted_result(cx: &LateContext<'_>, ty: ty::Ty) -> bool {
+    if let ty::Adt(adt, substs) = ty.kind() {
+        if !cx.tcx.is_diagnostic_item(sym::Result, adt.did()) {
+            return false;
+        }
+
+        // Don't lint if the error type is std::fmt::Error
+        if let Some(error_ty) = substs.get(1) {
+            if let ty::Adt(error_adt, _) = error_ty.expect_ty().kind() {
+                if match_def_path(cx, error_adt.did(), &["std", "fmt", "Error"]) {
```

Comment:
```suggestion
        if let Some(error_ty) = substs.get(1)
            && let ty::Adt(error_adt, _) = error_ty.expect_ty().kind()
            && match_def_path(cx, error_adt.did(), &["std", "fmt", "Error"]) {
```
Nit. Note some of the surrounding code will have to change.